File: notesApp/ios/DatabaseManager.swift
//
//  DatabaseManager.swift
//  notesApp
//
//  Created by Gonzalo Baldassarre on 17/08/2024.
//

import RealmSwift

class DatabaseManager {
  static let shared = DatabaseManager()
  private var realm: Realm?
  private let dateFormatter: ISO8601DateFormatter = {
    let formatter = ISO8601DateFormatter()
    formatter.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
    return formatter
  }()

  private init() {
    do {
      let config = Realm.Configuration(schemaVersion: 1, migrationBlock: nil, deleteRealmIfMigrationNeeded: true)
      Realm.Configuration.defaultConfiguration = config
      realm = try Realm()
    } catch {
      print("Error initializing Realm: \(error)")
    }
  }

  private func executeOnMainQueue<T>(_ operation: @escaping () -> T) -> T {
    if Thread.isMainThread {
      return operation()
    } else {
      var result: T!
      DispatchQueue.main.sync {
        result = operation()
      }
      return result
    }
  }
    
  func createNote(title: String, content: String, userId: String?, priority: String, _id: String? = nil, status: String?, isSynced: Bool?, createdAt: String?, editedAt: String?) -> String {
    executeOnMainQueue {
      var _createdAt: Date
      if let createdAtString = createdAt,
         let date = self.dateFormatter.date(from: createdAtString) {
        print("this date \(createdAtString) \(date)")
        _createdAt = date
      } else {
        print("no - date")
        _createdAt = Date()
      }
      
      var _editedAt: Date
      if let editedAtString = editedAt,
         let date = self.dateFormatter.date(from: editedAtString) {
        _editedAt = date
      } else {
        _editedAt = _createdAt
      }
      
      
      let note = Note()
      note._id = _id ?? "LocalId-\(UUID().uuidString)"
      note.title = title
      note.content = content
      note.createdAt = _createdAt
      note.editedAt = _editedAt
      note.userId = userId
      note.priority = priority
      note.isSynced = isSynced ?? true
      note.status = status ?? NoteStatus.ACTIVE.rawValue
      
      do {
        try self.realm?.write {
          self.realm?.add(note, update: .modified)
        }
      } catch {
        print("Error saving note: \(error)")
      }
      
      return note._id
    }
  }
    
  func updateNote(_id: String, title: String, content: String, priority: String, status: String, isSynced: Bool?) {
    executeOnMainQueue {
      guard let note = self.realm?.object(ofType: Note.self, forPrimaryKey: _id) else { return }
      
      do {
        try self.realm?.write {
          note.title = title
          note.content = content
          note.editedAt = Date()
          note.priority = priority
          note.status = status
          note.isSynced = isSynced ?? true
        }
      } catch {
        print("Error updating note: \(error)")
      }
    }
  }
  
  func updateNotes(notesData: [NSDictionary]) {
    executeOnMainQueue {
      print("Updating data in batch")
      do {
        try self.realm?.write {
          for noteData in notesData {
            guard let id = noteData["_id"] as? String,
                  let note = self.realm?.object(ofType: Note.self, forPrimaryKey: id) else {
              print("Error: No se encontró la nota con el id \(String(describing: noteData["id"]))")
              continue
            }
            note.title = noteData["title"] as? String ?? ""
            note.content = noteData["content"] as? String ?? ""
            note.editedAt = Date()
            note.priority = noteData["priority"] as? String ?? ""
            note.status = noteData["status"] as? String ?? ""
            note.isSynced = noteData["isSynced"] as? Bool ?? true
          }
        }
      } catch {
        print("Error updating notes: \(error)")
      }
    }
  }
    
  func removeNote(_id: String) {
    executeOnMainQueue {
      print("Busxcnado note")
      guard let note = self.realm?.object(ofType: Note.self, forPrimaryKey: _id) else { return }
      print("Encontre la note")
      do {
        try self.realm?.write {
          self.realm?.delete(note)
        }
      } catch {
        print("Error deleting note: \(error)")
      }
    }
  }

  func deleteNote(_id: String, isSynced: Bool?) {
    executeOnMainQueue {
      print("trying to delete \(_id)")
      guard let note = self.realm?.object(ofType: Note.self, forPrimaryKey: _id) else { return }
      do {
        try self.realm?.write {
          note.status = NoteStatus.DELETED.rawValue
          note.isSynced = isSynced ?? true
        }
      } catch {
        print("Error deleting note: \(error)")
      }
    }
  }
  
  func getNotes() -> [Note] {
    executeOnMainQueue {
      return self.realm?.objects(Note.self).map { $0 } ?? []
    }
  }

  func getUnsyncedNotes() -> [Note] {
    executeOnMainQueue {
      return self.realm?.objects(Note.self).filter("isSynced == false").map { $0 } ?? []
    }
  }
    
  func createNotes(notes: [NSDictionary]) -> [String] {
    executeOnMainQueue {
      var createdIds: [String] = []
      
      do {
        try self.realm?.write {
          for noteData in notes {
            var createdAt: Date
            var editedAt: Date

            if let createdAtString = noteData["createdAt"] as? String,
               let date = self.dateFormatter.date(from: createdAtString) {
              createdAt = date
            } else {
              createdAt = Date()
            }
            
            if let editedAtString = noteData["editedAt"] as? String,
               let date = self.dateFormatter.date(from: editedAtString) {
              editedAt = date
            } else {
              editedAt = createdAt
            }
            
            
            let note = Note()
            note._id = noteData["_id"] as? String ?? UUID().uuidString
            note.title = noteData["title"] as? String ?? ""
            note.content = noteData["content"] as? String ?? ""
            note.createdAt = createdAt
            note.editedAt = editedAt
            note.userId = noteData["userId"] as? String
            note.priority = noteData["priority"] as? String ?? ""
            note.isSynced = noteData["isSynced"] as? Bool ?? true
            note.status = noteData["status"] as? String ?? ""
            
            self.realm?.add(note, update: .modified)
            createdIds.append(note._id)
          }
        }
      } catch {
        print("Error creating notes in batch: \(error)")
      }
      
      return createdIds
    }
  }
}

File: notesApp/android/app/src/main/java/com/notesapp/database/DatabaseManager.kt
package com.notesapp.database
import io.realm.kotlin.Realm
import io.realm.kotlin.RealmConfiguration
import io.realm.kotlin.ext.query
import io.realm.kotlin.types.RealmObject
import io.realm.kotlin.types.annotations.PrimaryKey
import org.mongodb.kbson.BsonDateTime
import org.mongodb.kbson.ObjectId
import java.util.*

object DatabaseManager {
    private lateinit var realm: Realm

    init {
        val config = RealmConfiguration.create(schema = setOf(Note::class))
        realm = Realm.open(config)
    }

    fun createNote(title: String, content: String, userId: String?, priority: String, _id: String? = null, status: String?, isSynced: Boolean?, createdAt: Long?, editedAt: Long?): String {
        var noteId = ""
        try {
            realm.writeBlocking {
                val note = Note().apply {
                    this._id = _id ?: "LocalId-${UUID.randomUUID()}"
                    this.title = title
                    this.content = content
                    this.createdAt = createdAt ?: System.currentTimeMillis()
                    this.editedAt = editedAt ?: this.createdAt
                    this.userId = userId
                    this.priority = priority
                    this.isSynced = isSynced ?: true
                    this.status = status ?: NoteStatus.ACTIVE.name
                }
                copyToRealm(note)
                noteId = note._id
            }
        } catch (e: Exception) {
            throw RuntimeException("Error creating note: ${e.message}")
        }
        return noteId
    }

    fun updateNote(_id: String, title: String, content: String, priority: String, status: String, isSynced: Boolean?) {
        try {
            realm.writeBlocking {
                query<Note>("_id == $0", _id).first().find()?.let { note ->
                    note.title = title
                    note.content = content
                    note.editedAt = System.currentTimeMillis()
                    note.priority = priority
                    note.status = status
                    note.isSynced = isSynced ?: true
                }
            }
        } catch (e: Exception) {
            throw RuntimeException("Error updating note: ${e.message}")
        }
    }

    fun updateNotes(notesData: List<Map<String, Any?>>) {
        try {
            realm.writeBlocking {
                notesData.forEach { noteData ->
                    val id = noteData["_id"] as? String ?: return@forEach
                    query<Note>("_id == $0", id).first().find()?.let { note ->
                        note.title = noteData["title"] as? String ?: ""
                        note.content = noteData["content"] as? String ?: ""
                        note.editedAt = System.currentTimeMillis()
                        note.priority = noteData["priority"] as? String ?: ""
                        note.status = noteData["status"] as? String ?: ""
                        note.isSynced = noteData["isSynced"] as? Boolean ?: true
                    }
                }
            }
        } catch (e: Exception) {
            throw RuntimeException("Error updating notes: ${e.message}")
        }
    }

    fun removeNote(_id: String) {
        try {
            realm.writeBlocking {
                query<Note>("_id == $0", _id).first().find()?.let { delete(it) }
            }
        } catch (e: Exception) {
            throw RuntimeException("Error removing note: ${e.message}")
        }
    }

    fun deleteNote(_id: String, isSynced: Boolean?) {
        try {
            realm.writeBlocking {
                query<Note>("_id == $0", _id).first().find()?.let { note ->
                    note.status = NoteStatus.DELETED.name
                    note.isSynced = isSynced ?: true
                    note.editedAt = System.currentTimeMillis()
                }
            }
        } catch (e: Exception) {
            throw RuntimeException("Error deleting note: ${e.message}")
        }
    }

    fun getNotes(): List<Note> {
        return realm.query<Note>().find()
    }

    fun getUnsyncedNotes(): List<Note> {
        return realm.query<Note>("isSynced == false").find()
    }

    fun createNotes(notes: List<Map<String, Any?>>): List<String> {
        val createdIds = mutableListOf<String>()
        try {
            realm.writeBlocking {
                notes.forEach { noteData ->
                    val noteId = noteData["_id"] as? String ?: UUID.randomUUID().toString()
                    val existingNote = query<Note>("_id == $0", noteId).first().find()

                    val newEditedAt = noteData["editedAt"] as? Long ?: System.currentTimeMillis()

                    if (existingNote == null || newEditedAt > (existingNote?.editedAt ?: 0L)) {
                        val note = Note().apply {
                            _id = noteId
                            title = noteData["title"] as? String ?: ""
                            content = noteData["content"] as? String ?: ""
                            createdAt = (noteData["createdAt"] as? Long) ?: System.currentTimeMillis()
                            editedAt = newEditedAt
                            userId = noteData["userId"] as? String
                            priority = noteData["priority"] as? String ?: ""
                            isSynced = noteData["isSynced"] as? Boolean ?: true
                            status = noteData["status"] as? String ?: ""
                        }
                        copyToRealm(note)
                        createdIds.add(note._id)
                    }
                }
            }
        } catch (e: Exception) {
            throw RuntimeException("Error creating notes: ${e.message}")
        }
        return createdIds
    }

}